packages/openassistants/openassistants/eval/interaction.py

```diff
@@ -19,322 +14,232 @@
+        super().__init__(f"[{failure_type}] {comment}")
+        self.failure_type = failure_type
+        self.comment = comment
 
-class GradeSummary(BaseModel):
-    """Grade a summary against a data table."""
 
-    analysis: Annotated[
-        str,
-        Field(description="grading analysis of the summary"),
-    ]
-    pass_fail: Annotated[
-        Literal["pass", "fail"],
-        Field(description="does the summary pass or fail"),
-    ]
-
-    @staticmethod
-    def _remove_key_recursive(d: Any, target_key: str):
-        if isinstance(d, dict):
-            d.pop(target_key, "")
-            for key in d:
-                d[key] = GradeSummary._remove_key_recursive(d[key], target_key)
-        return d
-
-    @classmethod
-    def to_openai_function(cls) -> dict:
-        schema = convert_pydantic_to_openai_function(cls)  # type: ignore
-        if "description" in schema["parameters"]:
-            schema["parameters"].pop("description")
-        GradeSummary._remove_key_recursive(schema, "title")
-        return schema  # type: ignore
-
-
-InteractionTypes = Annotated[
-    Union["SentinelInteraction", "FunctionInteraction"],
-    Field(json_schema_extra={"discriminator": "type"}),
-]
+class BaseResponseChecker(abc.ABC):
+    @abc.abstractmethod
+    async def check(self, response: "FunctionInteractionResponse") -> None:
+        """
+        Check the response
+        :raises InteractionCheckError: if the check fails
+        """
+        pass
 
 
-class BaseInteraction(abc.ABC, BaseModel):
-    type: str
+class FunctionInteractionNode(BaseModel):
+    type: Literal["function"] = "function"
     message: str
-    children: List[InteractionTypes] = []
+    library: str
+    function: str
+    arg_schema: dict = {"type": "object"}
+    sample_args: Dict[str, Any] = {}
+    summarization_assertions: List[str] = []
 
-    @abc.abstractmethod
-    async def run(
-        self, chat_history: List[OpasMessage], assistant: Assistant
-    ) -> "InteractionReport":
-        pass
 
+class FunctionInteraction(FunctionInteractionNode):
+    children: List["FunctionInteraction"] = []
 
-class InteractionReport(BaseModel):
-    interaction: InteractionTypes
-    failures: List[str]
-    children: List["InteractionReport"]
-    function_response: Optional[OpasFunctionMessage] = None
-    summary_grading: Optional[GradeSummary] = None
-
-    @staticmethod
-    def cascade_failure(interaction: InteractionTypes, failure: str):
-        return InteractionReport(
-            interaction=interaction,
-            failures=[failure],
-            children=[
-                InteractionReport.cascade_failure(c, "cascading failure from parent")
-                for c in interaction.children
-            ],
+    async def run_function_selection(
+        self,
+        assistant: Assistant,
+        history_with_user_message: List[OpasMessage],
+    ) -> OpasAssistantMessage:
+        response_messages = await last_value(
```

Comment:
would be nice to add a type annotation here and on the other places in this file you set response_messages.